chain/consensus/common/crossmsg.go

```diff
@@ -0,0 +1,180 @@
+package common
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/lotus/blockstore"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/actors/builtin"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical/actors/sca"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical/subnet"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/vm"
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+	blockadt "github.com/filecoin-project/specs-actors/actors/util/adt"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"golang.org/x/xerrors"
+)
+
+func checkCrossMsg(pstore, snstore blockadt.Store, parentSCA, snSCA *sca.SCAState, msg *types.Message) error {
+	switch hierarchical.GetMsgType(msg) {
+	case hierarchical.Fund:
+		// sanity-check: the root chain doesn't support topDown messages,
+		// so return an error if parentSCA is nil and we are here.
+		if parentSCA == nil {
+			return xerrors.Errorf("root chains (id=%v) does not support topDown cross msgs", snSCA.NetworkName)
+		}
+		return checkTopDownMsg(pstore, parentSCA, snSCA, msg)
+	case hierarchical.Release:
+		panic("Not implemented")
+	case hierarchical.Cross:
+		panic("Not implemented")
+	}
+
+	return xerrors.Errorf("Unknown cross-msg type")
+}
+
+// checkTopDownMsg validates the topdown message.
```

Comment:
TopDown and DownTop is confusing, maybe TopDown and BottomUp?

Reply:
is top down from parent to child? then why can't the root send messages to the children?

Reply:
> TopDown and DownTop is confusing, maybe TopDown and BottomUp?

Great point. Will do a global renaming.

> Is top down from parent to child? then why can't the root send messages to the children?

Yes. Some transactions may not need to go all the way up to the root, addresses may not have the available balance in the root, and the SCA in the parent is the one tracking the available supply in the subnet. There may be a better way of handling all of this. I'll add all of this to the spec and we can discuss it further.